spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -316,20 +316,30 @@ public InternalRow read(Decoder decoder, Object reuse) throws IOException {
           break;
         }
       }
-      InternalRow struct = new GenericInternalRow(nullIndex >= 0 ? alts.size() - 1 : alts.size());
+
+      int index = decoder.readIndex();
+      if (index == nullIndex) {
+        // if it is a null data, directly return null as the whole union result
+        return readers[index].read(decoder, reuse);
+      }
+
+      // otherwise, we need to return an InternalRow as a struct data
+      InternalRow struct = new GenericInternalRow(nullIndex >= 0 ? alts.size() : alts.size() + 1);
       for (int i = 0; i < struct.numFields(); i += 1) {
         struct.setNullAt(i);
       }
 
-      int index = decoder.readIndex();
-      Object value = this.readers[index].read(decoder, reuse);
+      Object value = readers[index].read(decoder, reuse);
 
       if (nullIndex < 0) {
-        struct.update(index, value);
+        struct.update(index + 1, value);
+        struct.setInt(0, index);
       } else if (index < nullIndex) {
+        struct.update(index + 1, value);
+        struct.setInt(0, index);
+      } else {
```

Comment:
Not sure if I follow this, what does the relative position between value index and null index have anything to do with how we assign the value in the struct ?

Reply:
Because in avro the nullability of a union type is presented by the existence of a `NULL` type inside the union alternatives, and the `NULL` type can occur in any order inside the union, but in our converted struct, we don't have a field that corresponds to that `NULL` type, thus we have `-1` field count.

Reply:
I guess what I didn't know earlier was, if NULL type appears in the union it seems it has to be the first position in 
the alternative list ? But the avro spec said "Thus, for unions containing "null", the "null" is usually listed first," 

Will this lead to a silent failure if you don't check this first?

Reply:
I think that's just a recommendation from the Avro spec, not mandated, in fact, in our ecosystem there are many avro schemas which failed to put `null` as the first type element in the union.